proposals/0292-package-registry-service.md

```diff
@@ -38,77 +43,222 @@ it's not well-suited to this kind of workflow for the following reasons:
 is provided alongside this proposal.
 In addition,
-an OpenAPI (v3) document and a reference implementation
+an OpenAPI (v3) document
+and a reference implementation written in Swift
 are provided for the convenience of developers interested
 in building their own package registry.
 
 ### Changes to Swift Package Manager
 
+#### Package identity
+
+Currently, the identity of a package is computed from
+the last path component of its effective URL
+(which can be changed with dependency mirroring).
+However, this approach can lead to a conflation of
+distinct packages with similar names
+and the duplication of the same package under different names.
+
+We propose using a scoped identifier
+in the form `scope.package-name`
+to identify package dependencies.
+
+A *scope* provides a namespace for related packages within a package registry.
+A package scope consists of
+alphanumeric characters and hyphens.
+Hyphens may not occur at the beginning or end,
+nor consecutively within a scope.
+The maximum length of a package name is 39 characters.
+A valid package scope matches the following regular expression pattern:
+
+```regexp
+\A[a-zA-Z\d](?:[a-zA-Z\d]|-(?=[a-zA-Z\d])){0,39}\z
+```
+
+A package's *name* is specified by the `name` provided in its manifest.
+
+The maximum length of a package name is 128 characters.
+A valid package name matches the following regular expression pattern:
+
+```regexp
+\A\p{XID_Start}\p{XID_Continue}{0,127}\z
+```
+
+> For more information,
+> see [Unicode Identifier and Pattern Syntax][UAX31].
+
+Package scopes are case-insensitive
+(for example, `mona` ≍ `MONA`).
+Package names are
+case-insensitive,
+diacritic-insensitive
+(for example, `Å` ≍ `A`), and
+width-insensitive
+(for example, `Ａ` ≍ `A`).
+Package names are compared using
+[Normalization Form Compatible Composition (NFKC)][UAX15].
+
+#### New `PackageDescription` API
+
+The `Package.Dependency` type adds the following static method:
+
+```swift
+extension Package.Dependency {
+    /// Adds a dependency on a package with the specified identifier
+    /// that uses the provided version requirement.
+    public static func package(
+        id: String,
+        _ requirement: Package.Dependency.VersionBasedRequirement
+    ) -> Package.Dependency
+}
+```
+
+These methods may be called in the `dependencies` field of a package manifest
+to declare one or more dependencies by their respective package identifier.
+
+```swift
+dependencies: [
+   .package(id: "mona.LinkedList", .upToNextMinor(from: "1.1.0")),
+   .package(id: "mona.RegEx", .exact("2.0.0"))
+]
+```
+
+A package dependency declared with an identifier using this method
+may only specify a version-based requirement.
+`Package.Dependency.VersionBasedRequirement` is a new type
+that provides the same interface as `Package.Dependency.Requirement`
+for version-based requirements,
+but excluding branch-based and commit-based requirements.
+
+#### Module name collision resolution
```

Comment:
cc @elsh

Reply:
@mattt the title here is a tad confusing, since the focus is on deduping identities in a mixed graph (id + url)? 

in "Module name collision resolution" we usually refer to the work @elsh is exploring on using the package identity to drive a module prefix such that two packages that vend the same module name can co-exist. iow using a FQDN module name based on the package identity to prevent module name collisions